### PR TITLE
Add test for build tool depends with per-component builds disabled

### PR DESCRIPTION
**Pull request by alex-rozenshteyn-leapyear**
This seems to be the underlying issue in #6011

I'm not sure how to fix it (yet), but I'm hoping to solicit suggestions. My best guess is that the bug is in `Distribution.Client.ProjectPlanning`.

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.
* [ ] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import Test.Cabal.Prelude
+-- Test build-tool-depends between two packages
+main = cabalTest $ do
+    cabal "v2-build" ["client", "--disable-per-component"]
```
Comment (phadej) [inline cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs]: Even the test fails now, it will still fail as this flag says that `bad-do-not-build-me` have to be built too.

Will this test still fail as expected, if `bad-do-not-build-me` executable is removed?

File: cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import Test.Cabal.Prelude
+-- Test build-tool-depends between two packages
```
Comment (phadej) [inline cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs]: Could you name the test file `cabal.test.hs`, as it's testing `cabal-install` not `Setup`.

Also we have issue tests as e.g. `./PackageTests/Regression/T4202/cabal.test.hs` but I have no strong opinion on that.

File: cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs (hunk 1 above)
Comment (alex-rozenshteyn-leapyear) [reply cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs]: Yes. Removing the `bad-do-not-build-me` executable results in the following:

```
% cabal new-clean && cabal new-build client --disable-per-component
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - pre-proc-999.999.999 (exe:pre-proc) (first run)
 - client-0.1.0.0 (exe:hello-world) (first run)
Configuring pre-proc-999.999.999...
cabal: Unknown build target 'exe:pre-proc'.
There is no executable component 'pre-proc'.

cabal: Failed to build pre-proc-999.999.999 (which is required by
client-0.1.0.0).
```

I will make the change.

File: cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs (hunk 2 above)
Comment (alex-rozenshteyn-leapyear) [reply cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs]: I will rename the test file and move it to issue tests.

**Comment by alex-rozenshteyn-leapyear** [comment]
I'm a little stuck trying to fix the bug.

For some reason, with `--disable-per-component`, `exe:pre-proc` is passed to setup: 
```
Using self-exec internal setup method with build-type Simple and args:
["act-as-setup","--build-type=Simple","--","build","--verbose=3","--builddir=.../dist-newstyle/build/x86_64-linux/ghc-8.6.5/pre-proc-999.999.999","exe:pre-proc"]
```

It seems that [`external_cc_map`](https://github.com/haskell/cabal/blob/master/cabal-install/Distribution/Client/ProjectPlanning.hs#L1534-L1535) in `Distribution.Client.ProjectPlanning` has an entry of

```
( PackageName "pre-proc" 
, fromList 
    [ 
        ( CExeName ( UnqualComponentName "zero-to-one" )
        , AnnotatedId 
            { ann_pid = PackageIdentifier 
                { pkgName = PackageName "pre-proc" 
                , pkgVersion = mkVersion 
                    [ 999
                    , 999   
                    , 999   
                    ]       
                }       
            , ann_cname = CExeName ( UnqualComponentName "zero-to-one" )
            , ann_id = ComponentId "pre-proc-999.999.999-inplace" 
            }
        )       
    ]
) 
```

The `ComponentId` looks suspicious to me, but that could be a result of me not understanding what's supposed to be happening.

Comparing the `plan.json` files with and without `--disable-per-component` also might yield some insight.

Thoughts?

**Comment by phadej** [comment]
`-inplace` in component id  looks correct, it's "package-name-version-inplace" for "local" packages

**Comment by alex-rozenshteyn-leapyear** [comment]
> `-inplace` in component id looks correct, it's "package-name-version-inplace" for "local" packages

So how does the `build` step get its parameters? It seems to be getting an `exe:[component name]` parameter, which is correct for per-component builds, but not for per-package builds if the component name is the package name.

**Comment by alex-rozenshteyn-leapyear** [comment]
Found the source of the problem (turns out there was a TODO about it). I fixed it, but I'm not sure if the fix is the right way to do it. In a per-package build, I explicitly added all executables of a package to the build dependencies if any of the executables was a dependency.

**Comment by alex-rozenshteyn-leapyear** [comment]
Note that the single failure seems to be unrelated (https://travis-ci.org/haskell/cabal/jobs/606495569):
```
$ Installing caching utilities
/home/travis/.rvm/scripts/cli: line 872:  4372 Segmentation fault      "$rvm_scripts_path/list" "${rvm_ruby_args[@]}"
```

**Comment by alex-rozenshteyn-leapyear** [comment]
@phadej I've been dogfooding the this patch, and it seems to work. CI passed except for an unrelated failure. Is there anything else I need to do to get this merged?

**Comment by alex-rozenshteyn-leapyear** [comment]
Ping. Is there any action needed on my part to help merge this in?

**Comment by phadej** [comment]
Sorry, there were a bit of hassle in preparing the release branches. Could you rebase to the current master (the CI setup should be more robust there)

File: cabal-testsuite/PackageTests/Regression/T6011/pre-proc/pre-proc.cabal [hunk 3]
```diff
@@ -0,0 +1,12 @@
+name:                pre-proc
+version:             999.999.999
+synopsis:            Checks build-tool-depends are put in PATH
+license:             BSD3
+category:            Testing
+build-type:          Simple
```
Comment (phadej) [inline cabal-testsuite/PackageTests/Regression/T6011/pre-proc/pre-proc.cabal]: This is not `Custom`, so it's not what described in #6011

I'll check what happens with non-Custom builds, there we should be able to build just an executable (of possibly multiple in the package). I don't defaulting to build every exe, I'd expect someone to complain about that too.

File: cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs (hunk 1 above)
Comment (phadej) [reply cabal-testsuite/PackageTests/BuildToolDependsNonComponent/setup.test.hs]: Ok, now I see why there was second executable. We should not build it. Argh, this is tricky. I'll take this over to see how to solve this without compromises.

File: cabal-testsuite/PackageTests/Regression/T6011/cabal.test.hs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+import Test.Cabal.Prelude
+-- Test build-tool-depends between two packages
+main = cabalTest $ do
+    cabal "v2-build" ["client", "--disable-per-component"]
```
Comment (phadej) [inline cabal-testsuite/PackageTests/Regression/T6011/cabal.test.hs]: Looking at this, we probably should do something special in `ProjectPlanning` based on whether per-component build is disabled (either by flag or `build-type: Custom`). We **should not** default to do more in all scenarios just in case if per component build is disabled: everyone would pay for a corner-case.

File: cabal-testsuite/PackageTests/Regression/T6011/pre-proc/pre-proc.cabal (hunk 3 above)
Comment (alex-rozenshteyn-leapyear) [reply cabal-testsuite/PackageTests/Regression/T6011/pre-proc/pre-proc.cabal]: Correct. It's closely related but not a direct reproduction. It _appears_ to be the case that custom builds are not the core issue, but the fact that custom turns off per-component. Whether or not it effects custom builds, this is an issue.

File: cabal-testsuite/PackageTests/Regression/T6011/cabal.test.hs (hunk 4 above)
Comment (alex-rozenshteyn-leapyear) [reply cabal-testsuite/PackageTests/Regression/T6011/cabal.test.hs]: Could you elaborate on this? What would the "something special" look like? In what sense is everyone paying for the corner case in the current situation?

**Comment by alex-rozenshteyn-leapyear** [comment]
> Sorry, there were a bit of hassle in preparing the release branches. Could you rebase to the current master (the CI setup should be more robust there)

No worries; I understand release-related stresses. Since you're taking over the issue, I'll leave the rebasing to you to avoid conflicting?

**Comment by alex-rozenshteyn-leapyear** [comment]
@phadej Is any work being done on this?

**Comment by phadej** [comment]
Sorry, I have paged out the details already. I'll come back to this eventually.

**Comment by alex-rozenshteyn-leapyear** [comment]
@phadej No worries!

**Comment by Kleidukos** [comment]
Marking this PR as draft :slightly_smiling_face:

**Comment by ffaf1** [comment]
Hello, I am going through old PRs to check whether they are stale.

If this PR is still “live”, write a comment and I will remove the `consider closing` label. Otherwise in ≃ 2 weeks this PR will be closed.
